### PR TITLE
Add Pause for application + Router + Database

### PR DESCRIPTION
**Pull request by erebe**
Question for tomorrow: Should we scale_down router if they don't have any custom domain attached to it ?

- Added PAUSE action for application and Router
- For app: it scaledown the deployment to 0
- For Routers: it does nothing out of fear to be blacklisted by LetsEncrypt due to generating too much certificate
- Remove box::new(||) it does not do/add anything

### CODE REVIEW

File: src/cloud_provider/service.rs [hunk 1]
```diff
@@ -470,6 +470,41 @@ where
...
+        DeploymentTarget::SelfHosted(k, env) => {
+            (*k, *env)
+        },
+    };
+
+    let scaledown_ret = kubectl_exec_scale_replicas(
+        kubernetes.config_file_path()?,
+        kubernetes.cloud_provider().credentials_environment_variables(),
+        environment.namespace(),
+        ScalingKind::Deployment,
```
Comment (erebe) [inline src/cloud_provider/service.rs:496]: Going to FIX: I suppose I need to check if it is a statefulset or a deployment, as app can be both

File: tests/aws/aws_environment.rs [hunk 2]
```diff
@@ -106,6 +106,45 @@ fn deploy_a_working_environment_with_no_router_on_aws_eks() {
...
+            TransactionResult::Rollback(_) => assert!(false),
+            TransactionResult::UnrecoverableError(_, _) => assert!(false),
+        };
+
+        match ctx_pause_environment(&context_for_delete, &ea) {
+            TransactionResult::Ok => assert!(true),
+            TransactionResult::Rollback(_) => assert!(false),
+            TransactionResult::UnrecoverableError(_, _) => assert!(false),
+        };
+
```
Comment (deimosfr) [inline tests/aws/aws_environment.rs:138]: don't you add any check to validate the result?

File: src/cloud_provider/digitalocean/router.rs [hunk 3]
```diff
@@ -522,34 +522,24 @@ impl Create for Router {
...
+            || delete_stateless_service(target, self, true),
         )
     }
 }
 
 impl Pause for Router {
-    fn on_pause(&self, target: &DeploymentTarget) -> Result<(), EngineError> {
-        info!("DO.router.on_pause() called for {}", self.name());
-
-        send_progress_on_long_task(
```
Comment (deimosfr) [inline src/cloud_provider/digitalocean/router.rs:534]: why did you remove all send progress?

Comment (erebe) [reply src/cloud_provider/digitalocean/router.rs:534]: As discussed, we don't want to delete router when we send a PAUSE, we basically it just doing nothing, a NOP action

File: tests/aws/aws_environment.rs (hunk 2 above)
Comment (erebe) [reply tests/aws/aws_environment.rs:138]: it is the plan, when I get back the test cluster

File: src/cloud_provider/aws/external_service.rs [hunk 4]
```diff
@@ -199,49 +197,38 @@ impl Create for ExternalService {
...
-    fn on_pause_error(&self, target: &DeploymentTarget) -> Result<(), EngineError> {
+    fn on_pause_error(&self, _target: &DeploymentTarget) -> Result<(), EngineError> {
         warn!("AWS.external_service.on_pause_error() called for {}", self.name());
-
-        send_progress_on_long_task(
-            self,
-            crate::cloud_provider::service::Action::Pause,
-            Box::new(|| delete_stateless_service(target, self, true)),
-        )
+        Ok(())
```
Comment (erebe) [inline src/cloud_provider/aws/external_service.rs:221]: going to fix

File: src/cloud_provider/aws/databases/mongodb.rs [hunk 5]
```diff
@@ -259,37 +258,30 @@ impl Create for MongoDB {
...
     }
 
-    fn on_pause_error(&self, _target: &DeploymentTarget) -> Result<(), EngineError> {
+    fn on_pause_error(&self, target: &DeploymentTarget) -> Result<(), EngineError> {
         warn!("AWS.MongoDB.on_pause_error() called for {}", self.name());
 
-        // TODO what to do if there is a pause error?
-
-        Ok(())
+        self.on_pause(target)
```
Comment (deimosfr) [inline src/cloud_provider/aws/databases/mongodb.rs:276]: doesn't it create an infinite loop?

File: src/cloud_provider/digitalocean/application.rs [hunk 6]
```diff
@@ -299,23 +303,24 @@ impl Create for Application {
...
-            self,
-            crate::cloud_provider::service::Action::Pause,
-            Box::new(|| delete_stateless_service(target, self, false)),
-        )
+        send_progress_on_long_task(self, crate::cloud_provider::service::Action::Pause, || {
+            scale_down_service(
+                target,
+                self,
+                0,
+                if self.is_stateful() { Statefulset } else { Deployment },
```
Comment (deimosfr) [inline src/cloud_provider/digitalocean/application.rs:321]: nice catch

File: tests/aws/aws_databases.rs [hunk 7]
```diff
@@ -61,6 +61,49 @@ fn deploy_an_environment_with_3_databases_and_3_apps() {
...
+        TransactionResult::Rollback(_) => assert!(false),
+        TransactionResult::UnrecoverableError(_, _) => assert!(false),
+    };
+
+    // Check that we have actually 0 pods running for this db
+    let app_name = format!("postgresql{}-0", environment.databases[0].name);
+    let ret = get_pods_aws(environment.clone(), app_name.clone().as_str(), secrets.clone());
+    assert_eq!(ret.is_ok(), true);
+    assert_eq!(ret.unwrap().items.is_empty(), true);
+
```
Comment (deimosfr) [inline tests/aws/aws_databases.rs:99]: could you please add a test for the resume as well, to be 100% covered

File: src/cloud_provider/aws/databases/mongodb.rs (hunk 5 above)
Comment (erebe) [reply src/cloud_provider/aws/databases/mongodb.rs:276]: Going to remove it.
Think it just going to do it twice. It just that it is not very consistent between action & provider, some time we just redo in the xxx_err the same thing than in the exec

File: src/cloud_provider/aws/databases/mongodb.rs [hunk 8]
```diff
@@ -5,8 +5,9 @@ use tera::Context as TeraContext;
 use crate::cloud_provider::environment::Kind;
 use crate::cloud_provider::service::{
     check_service_version, default_tera_context, delete_stateful_service, deploy_stateful_service, get_tfstate_name,
-    get_tfstate_suffix, send_progress_on_long_task, Action, Backup, Create, Database, DatabaseOptions, DatabaseType,
-    Delete, Downgrade, Helm, Pause, Service, ServiceType, StatefulService, Terraform, Upgrade,
+    get_tfstate_suffix, scale_down_database, send_progress_on_long_task, Action, Backup, Create, Database,
+    DatabaseOptions, DatabaseType, Delete, Downgrade, Helm, Pause, Service, ServiceType, StatefulService, Terraform,
+    Upgrade,
 };
```
Comment (benjaminch) [inline src/cloud_provider/aws/databases/mongodb.rs:11]: OPT: question, maybe it starts to be a lot of individual imports? I am not sure how bad would importing `use crate::cloud_provider::service;` and use `service::get_tfstate_suffix` for example.
That's really a question, I have no idea if there are any downside.

File: src/cloud_provider/aws/databases/mysql.rs [hunk 9]
```diff
@@ -6,8 +6,9 @@ use crate::cloud_provider::aws::databases::utilities::{get_parameter_group_from_
 use crate::cloud_provider::environment::Kind;
 use crate::cloud_provider::service::{
     check_service_version, default_tera_context, delete_stateful_service, deploy_stateful_service, get_tfstate_name,
```
Comment (benjaminch) [inline src/cloud_provider/aws/databases/mysql.rs:8]: OPT: same regarding usings

File: src/cloud_provider/aws/databases/postgresql.rs [hunk 10]
```diff
@@ -6,8 +6,9 @@ use crate::cloud_provider::aws::databases::utilities::rds_name_sanitizer;
 use crate::cloud_provider::environment::Kind;
 use crate::cloud_provider::service::{
     check_service_version, default_tera_context, delete_stateful_service, deploy_stateful_service, get_tfstate_name,
```
Comment (benjaminch) [inline src/cloud_provider/aws/databases/postgresql.rs:8]: OPT: same

File: src/cloud_provider/aws/databases/redis.rs [hunk 11]
```diff
@@ -5,8 +5,9 @@ use tera::Context as TeraContext;
 use crate::cloud_provider::environment::Kind;
 use crate::cloud_provider::service::{
     check_service_version, default_tera_context, delete_stateful_service, deploy_stateful_service, get_tfstate_name,
```
Comment (benjaminch) [inline src/cloud_provider/aws/databases/redis.rs:7]: OPT: same

File: src/cloud_provider/service.rs [hunk 12]
```diff
@@ -470,6 +471,68 @@ where
...
+        DeploymentTarget::SelfHosted(k, env) => (*k, *env),
+    };
+
+    let scaledown_ret = kubectl_exec_scale_replicas(
+        kubernetes.config_file_path()?,
+        kubernetes.cloud_provider().credentials_environment_variables(),
+        environment.namespace(),
+        Statefulset,
+        service.sanitized_name().as_str(),
+        replicas_count as u32,
```
Comment (benjaminch) [inline src/cloud_provider/service.rs:493]: Even if not likely to happen, maybe worth being more cautious one this conversion?
What about using `usize::try_from(replicas_count).unwrap();`?

File: src/cloud_provider/service.rs [hunk 13]
```diff
@@ -470,6 +471,68 @@ where
...
+        }
+        DeploymentTarget::SelfHosted(k, env) => (*k, *env),
+    };
+    let scaledown_ret = kubectl_exec_scale_replicas(
+        kubernetes.config_file_path()?,
+        kubernetes.cloud_provider().credentials_environment_variables(),
+        environment.namespace(),
+        scaling_kind,
+        service.sanitized_name().as_str(),
+        replicas_count as u32,
```
Comment (benjaminch) [inline src/cloud_provider/service.rs:526]: Even if not likely to happen, maybe worth being more cautious one this conversion?
What about using `usize::try_from(replicas_count).unwrap();`?
